src/Assignments/SimpleProblemPage.tsx

```diff
@@ -6,98 +6,305 @@ import ProblemIframe from './ProblemIframe';
+                setVersionsRemaining(topic.topicAssessmentInfo.maxVersions - topic.topicAssessmentInfo.studentTopicAssessmentInfo.length);
+                setVersionId(topic.topicAssessmentInfo.studentTopicAssessmentInfo[0].id!);
+                setTopic(topic);
+            }
+        } else if (topic.topicTypeId === 2 && !_.isNil(topic.topicAssessmentInfo)) { // we are definitely an assessment - topicAssessmentInfo *should* never be missing
+            const usedVersions = topic.topicAssessmentInfo.studentTopicAssessmentInfo?.length ?? 0;
+            const actualVersionsRemaining = topic.topicAssessmentInfo.maxVersions - usedVersions;
+            setVersionsRemaining(actualVersionsRemaining);
+            if (actualVersionsRemaining > 0) {
+                confirmStartNewVersion(topic, actualVersionsRemaining, res.data.message);
+            } else {
+                // no problems were sent back, and user has used the maximum versions allowed
+                setError(`${res.data.message} You have used all available versions for this assessment.`);
+            }
+        }
+    };
 
     // This should always be used on the selectedProblem.
     const setProblemStudentGrade = (val: any) => {
         if (_.isEmpty(problems) || problems === null || _.isNaN(selectedProblemId) || selectedProblemId === null) return;
         problems[selectedProblemId].grades = [val];
-        setProblems({...problems});
+        setProblems({ ...problems });
+    };
+
+    const confirmSubmitVersion = (topicId: number, versionId: number) => {
+        setConfirmationParameters({
+            show: true,
+            headerContent: <h5>Submit my exam</h5>,
+            bodyContent: 'You are about to use one of your graded submissions. Proceed?',
+            onConfirm: async () => await getResultsOfSubmission(topicId, versionId),
+            onHide: () => setConfirmationParameters(DEFAULT_CONFIRMATION_PARAMETERS),
+        });
+    };
+
+    const confirmStartNewVersion = (topic: TopicObject, actualVersionsRemaining?: number, message?: string) => {
+        actualVersionsRemaining = actualVersionsRemaining ?? versionsRemaining; // retrieve from state if not supplied
+        setConfirmationParameters({
+            show: true,
+            headerContent: <h5>Begin a new version</h5>,
+            bodyContent: <div>
+                {message} <br />
+                You have {actualVersionsRemaining} {(actualVersionsRemaining === 1) ? ' version ' : ' versions '} remaining.<br />
+                Are you ready to begin a new version of this assessment?
+            </div>,
+            onHide: () => setConfirmationParameters(DEFAULT_CONFIRMATION_PARAMETERS),
+            onConfirm: async () => await loadNewVersion(topic)
+        });
+    };
+
+    const confirmEndVersion = () => {
+        let message = '';
+        if (!_.isNil(attemptsRemaining) && attemptsRemaining > 0) {
+            const nit = (attemptsRemaining === 1) ? 'attempt' : 'attempts';
```

Comment:
🤣 nice

You could just do it with the s itself, you don't have to do the whole word, either way, is fine it doesn't matter
This way is probably more readable

Reply:
😝